text/0000-remove-stacks-mixins.md

```diff
@@ -0,0 +1,128 @@
+# Meta
+[meta]: #meta
+- Name: Remove Stacks & Mixins
+- Start Date: 2021-06-30
+- Author(s): sclevine
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: [RFC0069](https://github.com/buildpacks/rfcs/blob/main/text/0069-stack-buildpacks.md), [RFC#167](https://github.com/buildpacks/rfcs/pull/167), many others
+
+# Summary
+[summary]: #summary
+
+*NOTE: This proposal is part of a larger initiative to reduce complexity originally outlined in https://github.com/buildpacks/rfcs/pull/167*
+
+This RFC proposes that we remove the "stack" and "mixin" concepts from the project and replace them with existing constructs in the container image ecosystem such as base images, Dockerfiles, and OS packages.
+
+# Motivation
+[motivation]: #motivation
+
+The "stack" and "mixin" concepts add unnecessary complexity to the project and make it difficult for new users and contributors to understand how buildpacks work. Compatibility guarantees that are strongly enforced by the stack contract could be replaced with metadata validations and warnings.
+
+# What it is
+[what-it-is]: #what-it-is
+
+Summary of changes:
+- Remove mixins
+- Replace stack metadata (including stack IDs) with canonical OS metadata.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+## Base Image Metadata
+
+Instead of a stack ID, runtime and build-time base images are labeled with the following canonicalized metadata:
+- OS (e.g., "linux", `$GOOS`)
+- Architecture (e.g., "x86_64", `$GOARCH`)
+- Distribution (optional) (e.g., "ubuntu", `$ID`)
+- Version (optional) (e.g., "18.04", `$VERSION_ID`)
+
+OS and Architecture must be valid identifiers as defined [here](https://golang.org/doc/install/source#environment).
+For Linux-based images, each field should be canonicalized against values specified in `/etc/os-release` (`$ID` and `$VERSION_ID`).
+
+The `stacks` list in `buildpack.toml` is replaced by a `targets` list, where each entry corresponds to a different buildpack image that is exported into a [manifest index](https://github.com/opencontainers/image-spec/blob/master/image-index.md). Each entry may contain multiple valid values for Distribution and/or Version, but only a single OS and Architecture.
```

Comment:
Would we support a wildcards in the `target`?

Reply:
Is the `target` mandatory for buildpacks. What about a very universal buildpack compatible with all supported images?

Reply:
Does the omission of `[targets]` in `buildpack.toml` mean `* <any stack>`?

Reply:
At least one OS/arch combination is necessary, but it doesn't need to include a distro/version.

I'm hesitant to make OS/arch (e.g., linux x86) optional, because OCI images must contain these fields.

Reply:
If we were to support wildcards, could the builder or platform provide the OS/arch labels for the image? Surely it can infer this from the runtime base image being targeted.

Reply:
I just meant for the buildpack image. If we supported a single buildpack image that could run on multiple architectures, it's `architecture` field would incorrect. Do you have use cases for multi-OS or multi-architecture buildpacks? The correct OS/architecture will get selected automatically when the buildpack image is pulled.

Reply:
My goal here is to eliminate boiler plate, and to make buildpack authorship simpler. Not necessarily to support multiple archs

Reply:
With `stack` these details where encapsulated and hidden behind IDs, while the builder owner managed what those IDs mapped to. By removing stack, we're pushing that onto the buildpack author (who now need to ensure that the `targets` they put in their `buildpack.toml` align with all the targets in the other buildpacks they want to combine with.

Reply:
Actually, I think there are plenty of multi-OS or multi-architecture buildpacks. Maybe Maven, Gradle, etc. Most buildpacks that run something atop a language VM. Maybe https://github.com/jkutner/no-secrets-buildpack, which just scans a repos for secrets (doesn't even provide or install anything)

Reply:
Added default values (matching OS & x86_64) to reduce boilerplate. Now buildpack authors only need to care about targets when they want to restrict usage of their buildpack to a specific distro/version, or when they need support for less common architectures. The OS is detected automatically (`/bin/build` vs. `/bin/build.bat`).

OCI images must have a single OS and single architecture field, and I don't think we should create buildpack images for more than one architecture automatically. Can we iterate on this in the future?